src/log_surgeon/finite_automata/Dfa.hpp

```diff
@@ -28,95 +49,404 @@ class Dfa {
+     * config in `rhs`. A config is considered mapped if both contain the same start, history, and
+     * registers.
+     *
+     * @param lhs The first set of configurations.
+     * @param rhs The second set of configurations.
+     * @return The register mapping if a bijection is possible.
+     * @return std::nullopt otherwise.
+     */
+    [[nodiscard]] static auto try_get_mapping(
+            ConfigurationSet const& lhs,
+            ConfigurationSet const& rhs
+    ) -> std::optional<std::unordered_map<reg_id_t, reg_id_t>>;
+
+    /**
+     * Creates a DFA state based on the given config set if the config does not already exist and
+     * cannot be mapped to an existing config. In the case of a new DFA state, it is added to
+     * `m_states`, `dfa_states`, and `unexplored_sets`.
+     *
+     * @param config_set The configuration set for which to create or get the DFA state.
+     * @param dfa_states Returns an updated map of configuration sets to DFA states.
+     * @param unexplored_sets Returns a queue of unexplored states.
+     * @return If `new_config_set` is already in `dfa_states`, a pair containing:
+     * - The existing DFA state.
+     * - std::nullopt.
+     * @return If `new_config_set` can be mapped to an existing config in `dfa_states`, a pair
+     * containing:
+     * - The existing DFA state.
+     * - The register mapping.
+     * @return Otherwise, a pair containing:
+     * - The newly created DFA state.
+     * - std::nullopt.
+     */
+    auto create_or_get_dfa_state(
+            ConfigurationSet const& config_set,
+            std::map<ConfigurationSet, TypedDfaState*>& dfa_states,
+            std::queue<ConfigurationSet>& unexplored_sets
+    ) -> std::pair<TypedDfaState*, std::optional<std::unordered_map<reg_id_t, reg_id_t>>>;
+
+    /**
+     * Determines the out-going transitions from the configuration set based on its NFA states.
+     *
+     * @param num_tags Number of tags in the NFA.
+     * @param config_set The configuration set.
+     * @param tag_id_with_op_to_reg_id Returns an updated mapping from operation tag id to
+     * register id.
+     * @return A map of input symbols to transitions. Each transition contains a vector of register
+     * operations and a destination configuration set.
+     */
+    [[nodiscard]] auto get_transitions(
+            size_t num_tags,
+            ConfigurationSet const& config_set,
+            std::map<tag_id_t, reg_id_t>& tag_id_with_op_to_reg_id
+    ) -> std::map<uint32_t, std::pair<std::vector<RegisterOperation>, ConfigurationSet>>;
+
+    /**
+     * Iterates over the configurations in the closure to:
+     * - Add the new registers needed to track the tags to 'm_reg_handler'.
+     * - Determine the operations to perform on the new registers.
+     *
+     * @param num_tags Number of tags in the NFA.
+     * @param closure Returns the set of DFA configurations with updated `tag_to_reg_ids`.
```

Comment:
```suggestion
     * @param closure Returns the set of DFA configurations with updated `tag_id_to_reg_ids`.
```
Trying to make things consistent since we tend to use `tag_id` in mappings.